src/TestFramework/TestFramework/Assertions/CollectionAssert.cs

```diff
@@ -1605,43 +1605,69 @@ internal static bool IsSubsetOfHelper(ICollection subset, ICollection superset)
         ref string reason)
     {
         Assert.CheckParameterNotNull(comparer, "Assert.AreCollectionsEqual", "comparer", string.Empty);
-        if (!ReferenceEquals(expected, actual))
+        if (ReferenceEquals(expected, actual))
         {
-            if ((expected == null) || (actual == null))
-            {
-                return false;
-            }
+            reason = string.Format(CultureInfo.CurrentCulture, FrameworkMessages.BothCollectionsSameReference, string.Empty);
+            return true;
+        }
 
-            if (expected.Count != actual.Count)
-            {
-                reason = FrameworkMessages.NumberOfElementsDiff;
-                return false;
-            }
+        return CompareIEnumerable(expected, actual, comparer, ref reason);
+    }
+
+    private static bool CompareIEnumerable(IEnumerable? expected, IEnumerable? actual, IComparer comparer, ref string reason)
+    {
+        if ((expected == null) || (actual == null))
+        {
+            return false;
+        }
 
-            IEnumerator expectedEnum = expected.GetEnumerator();
-            IEnumerator actualEnum = actual.GetEnumerator();
-            int i = 0;
-            while (expectedEnum.MoveNext() && actualEnum.MoveNext())
+        var stack = new Stack<Tuple<IEnumerator, IEnumerator, int>>();
+        stack.Push(new(expected.GetEnumerator(), actual.GetEnumerator(), 0));
+
+        while (stack.Count > 0)
+        {
+            Tuple<IEnumerator, IEnumerator, int> cur = stack.Pop();
+            IEnumerator expectedEnum = cur.Item1;
+            IEnumerator actualEnum = cur.Item2;
+            int position = cur.Item3;
+
+            while (expectedEnum.MoveNext())
             {
-                bool areEqual = comparer.Compare(expectedEnum.Current, actualEnum.Current) == 0;
-                if (!areEqual)
+                if (!actualEnum.MoveNext())
+                {
+                    reason = FrameworkMessages.NumberOfElementsDiff;
+                    return false;
+                }
+
+                object? curExpected = expectedEnum.Current;
+                object? curActual = actualEnum.Current;
+
+                if (curExpected is IEnumerable curExpectedEnum && curActual is IEnumerable curActualEnum)
```

Comment:
Do we have a test with a collection type that isn't `ICollection` to avoid regression?

Reply:
I did a test with List<ReadOnlyCollection<int>>

Reply:
Yes sorry I am blind...